website/blog/2020-05-28-summing-root-to-leaf-numbers.md

```diff
@@ -0,0 +1,127 @@
+            # Leaf
+            if not node.left and not node.right:
+                return partial_sum
+            # Non Leaf 
+            return (sum_root_to_leaf_helper(node.left, partial_sum) + sum_root_to_leaf_helper(node.right, partial_sum))
+```
+
+We use a default value for the partial sum to be 0. 
+
+In our main method, we want to include the `sum_root_to_leaf_helper` method as a nested method and simply pass on the node parameter. Finally, this is how our solution looks:
+
+```
+class Solution: 
+    def sumNumbers(self, root: TreeNode) -> int:
+        def sum_root_to_leaf_helper(node, partial_sum=0):
+            if not node:
+                return 0
+            partial_sum = partial_sum * 10 + node.val
+            # Leaf
+            if not node.left and not node.right:
+                return partial_sum
+            # Non Leaf 
+            return (sum_root_to_leaf_helper(node.left, partial_sum) + sum_root_to_leaf_helper(node.right, partial_sum))
+    
+        return sum_root_to_leaf_helper(root)
+```
+
+## The Algorithmic Complexity 
+
+When we come up with a solution, it is important to analyze its algorithmic complexity not only to estimate its performance but also to identify areas for improvement and reflect on our problem-solving skills. We should always ask the question: *can we do better than X?* Where X is the current complexity of our solution. 
+
+Time:
+
+Our solution is a modification of the depth-first-search pre-order traversal where we visit all nodes exactly once and perform a trivial computation (moving digits by integer multiplication). Thus, our runtime is simply `O(N)` where `N` represents the number of nodes in the given tree. A solution better than `O(N)` doesn't seem possible because to construct a number from digits, we need to know all the digits (and thus visit all nodes).
+
+Space:
+
+In terms of storage, we incur a high cost in the recursion call stack that builds up as our `sum_root_to_leaf_helper` calls itself. These calls *build-up* as one waits for another to finish.
+
+The maximum call stack is dependent upon the height of the binary tree (since we start backtracking after we visit a leaf), giving a complexity of `O(H)` where `H` is the height of the binary tree. In the worst case, the binary tree is skewed in either direction and thus `H = N`. Therefore, the worst-case space complexity is `O(N)`.
+
+You can read [this article](https://www.freecodecamp.org/news/how-recursion-works-explained-with-flowcharts-and-a-video-de61f40cb7f9/) to know more about recursion call stacks.
+
+It is possible to do better than `O(N)` by using a Morris Preorder Traversal. The basic idea is to link a node and its predecessor temporarily. You can read more about it [here](https://www.sciencedirect.com/science/article/abs/pii/0020019079900681).
+
+## The Conclusion 
+
+I hope this post helped! Please do let me know if you have any feedback, comments or suggestions by responding to this post.
```

Comment:
I don't have a commenting system. Do you have a Twitter account or any form of contact?

Reply:
Yes I do! [RaivatShah](https://twitter.com/RaivatShah)